### ISSUE TITLE
'history.state' object is getting lost when navigate using back/forward buttons

### ISSUE THREAD

**Issue by alecvolo** [angular/angular#28108]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/router....

### Is this a regression?

 No

### Description

I think there is a small bug - the `history.state` object is getting lost when navigate using back/forward buttons.
The fix should be pretty simple- set the restored state into the `NavigationExtras`  parameter in the call of `scheduleNavigation` in `Router.setUpLocationChangeListener()` method :

```javascript
        // Navigations coming from Angular router have a navigationId state property. When this
        // exists, restore the state.
        const state = change.state && change.state.navigationId ? change.state : null;
     setTimeout(
            () => { this.scheduleNavigation(rawUrlTree, source, state, {replaceUrl: true, state }); }, 0);
```

## 🔬 Minimal Reproduction

[https://stackblitz.com/edit/angular-bupuzn](https://stackblitz.com/edit/angular-bupuzn) created by @fjauer - just use back/forward after pressing Test button

## 🔥 Exception or Error
<pre><code>

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 7.2.1
Node: 10.15.0
OS: win32 x64
Angular: 7.2.0
... animations, common, compiler, compiler-cli, core, forms
... http, language-service, platform-browser
... platform-browser-dynamic, router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.12.1
@angular-devkit/build-angular     0.12.1
@angular-devkit/build-optimizer   0.12.1
@angular-devkit/build-webpack     0.12.1
@angular-devkit/core              7.2.1
@angular-devkit/schematics        7.2.1
@angular/cdk                      7.2.1
@angular/cli                      7.2.1
@angular/flex-layout              7.0.0-beta.23
@angular/material                 7.2.1
@ngtools/webpack                  7.2.1
@schematics/angular               7.2.1
@schematics/update                0.12.1
rxjs                              6.0.0
typescript                        3.2.2
webpack                           4.23.1

</code></pre>

**Anything else relevant?**

**Comment by alecvolo** [comment]
The workaround is to catch `NavigationStart` event and update ` CurrentNavigation` object
```javascript
export class AppComponent  {
  constructor(private _router: Router) {
    this._router.events.pipe(
        filter(e =>e instanceof NavigationStart),
        filter((e: NavigationStart)=>e.navigationTrigger == "popstate")
    )
      .subscribe((x: NavigationStart) => {
        this._router.getCurrentNavigation().extras.state={...x.restoredState, navigationId:x.id } ;
      });
  }
  }
```
Here is the StackBlitz example [https://stackblitz.com/edit/angular-xyrmg1](https://stackblitz.com/edit/angular-xyrmg1)  (Should be run in new window)

**Comment by jasonaden** [comment]
Thanks for the issue. Is this something you think you can create a PR for? I would like to get it fixed ASAP, but it will be a little while until this can be pulled in.

**Comment by alecvolo** [comment]
I've created PR  https://github.com/angular/angular/pull/28176

**Comment by Airblader** [comment]
@jasonaden Are we sure this is a bug? I would argue that this is a new navigation triggered by the user, so why should the state be kept here?

In fact, for me this would be a problem. I am listening for changes to query parameters, but need to ignore these updates when the navigation has been triggered by a specific part of my code. I currently solve this by passing a flag via the state when I trigger the navigation.

If the user clicks the back button, this is now a user-triggered navigation, so passing the state would lead to an incorrect result.

**Comment by Airblader** [comment]
Ah, I see that's just how pushstate works… that's most unfortunate (for me, that is). I guess this fix does make sense then. I guess I can also check that `trigger === "imperative"` for my case, perhaps.

**Comment by nishadabdulazeez** [comment]
Is this change now available in v7.2.9?
